sumpy/codegen.py

```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
 
 # }}}
 
 
-class MathConstantRewriter(CSECachingMapperMixin, IdentityMapper):
-    def map_variable(self, expr):
+class MathConstantRewriter(CSECachingMapperMixin, CallExternalRecMapper):
+    def map_variable(self, expr, *args):
         if expr.name == "pi":
             return prim.Variable("M_PI")
         else:
-            return IdentityMapper.map_variable(self, expr)
+            return expr
 
     map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
 
 
+def combine_mappers(*mappers):
+    """Returns a mapper that combines the work of several mappers.
+    For this to work, the mappers need to be instances of
+    :class:`sumpy.codegen.CallExternalRecMapper` and when calling
+    parent class methods, the mappers need to use the argument
+    *rec_object* passed to the *map_* method. All the mappers
+    need to be commutative and for any mapper applying the
+    mapper on an expression that has already been mapped should
+    result in an identical object.
+    """
+    from collections import defaultdict
+    all_methods = defaultdict(list)
+    base_classes = [CSECachingMapperMixin, IdentityMapper]
+    for mapper in mappers:
+        assert isinstance(mapper, CallExternalRecMapper)
+        for method_name in dir(type(mapper)):
+            if not method_name.startswith("map"):
+                continue
+            if method_name == "map_common_subexpression_uncached":
+                continue
+            method = getattr(type(mapper), method_name)
+            method_equals_base_class_method = False
+            for base_class in base_classes:
+                base_class_method = getattr(base_class, method_name, None)
+                if base_class_method is not None:
+                    method_equals_base_class_method = (base_class_method == method)
+                    break
+            else:
+                raise RuntimeError(f"Unknown mapping method {method_name}")
+
+            if method_equals_base_class_method:
+                continue
+            all_methods[method_name].append((mapper, method))
+
+    class CombineMapper(CSECachingMapperMixin, IdentityMapper):
+        def __init__(self, all_methods):
+            self.all_methods = all_methods
+        map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
+
+    def _map(method_name, self, expr, rec_object=None, *args):
+        if method_name not in self.all_methods:
+            return getattr(IdentityMapper, method_name)(self, expr)
+        for mapper, method in self.all_methods[method_name]:
+            new_expr = method(mapper, expr, self)
+            if new_expr is not expr:
+                return self.rec(new_expr)
+        return expr
+
+    from functools import partial
+    import types
+    combine_mapper = CombineMapper(all_methods)
+    for method_name in all_methods.keys():
+        setattr(combine_mapper, method_name,
+                types.MethodType(partial(_map, method_name), combine_mapper))
```

Comment:
Interesting, TIL! (Could've done that in https://github.com/inducer/pytools/pull/76/files, cc @alexfikl)